File: Algo/Leetcode/229MajorElementII.cpp
// advanced moore
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int count1 = 0; int val1 = 0;
        int count2 = 0; int val2 = 0;
        for(auto n:nums){
            if(count1 > 0 && val1 == n) count1++;
            else if(count2 > 0 && val2 == n) count2++;
            else if(count1==0) {val1 = n; count1=1;}
            else if(count2==0) {val2 = n; count2=1;}
            else{count1--; count2--;}
        }

        count1 = 0;
        count2 = 0;
        for(auto n:nums){
            if(n==val1) count1++;
            if(n==val2) count2++;
        }

        vector<int> res;
        if(count1>nums.size()/3) res.push_back(val1);
        if(count2>nums.size()/3 && val1!=val2) res.push_back(val2);
        return res;
    }
};

File: Algo/Leetcode/229MajorElementII.py
# Use dict
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        m = {}
        for n in nums:
            if m.get(n) is not None:
                m[n] += 1
            else:
                m[n] = 1

        res = []
        for k, v in m.items():
            if v > len(nums) / 3:
                res.append(k)

        return res

# Use moore
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        count1, count2 = 0, 0
        elem1, elem2 = None, None
        # Make sure that each elem is identical
        for n in nums:
            if (n == elem1):
                count1 += 1
            elif (n == elem2):
                count2 += 1
            elif count1 == 0:
                count1 += 1
                elem1 = n
            elif count2 == 0:
                count2 += 1
                elem2 = n
            else:
                count1 -= 1
                count2 -= 1

        # recount the num
        count1, count2 = 0, 0
        for n in nums:
            if n == elem1:
                count1 += 1
            if n == elem2:
                count2 += 1

        # count super num
        res = []
        if count1 > len(nums) / 3:
            res.append(elem1)
        if count2 > len(nums) / 3:
            res.append(elem2)

        return res